compiler/tests/fixtures/demos/uniswap.fe

```diff
@@ -0,0 +1,365 @@
+        else:
+            liquidity = self.min((amount0 * total_supply) / reserve0, (amount1 * total_supply) / reserve1)
+
+        assert liquidity > 0, "UniswapV2: INSUFFICIENT_LIQUIDITY_MINTED"
+
+        self._mint(to, liquidity)
+        self._update(balance0, balance1, reserve0, reserve1)
+
+        if fee_on:
+            self.k_last = reserve0 * reserve1 # reserve0 and reserve1 are up-to-date
+
+        emit Mint(msg.sender, amount0, amount1)
+        return liquidity
+
+    # this low-level function should be called from a contract which performs important safety checks
+    # TODO: support structs in the public interface (https://github.com/ethereum/fe/issues/282)
+    pub def burn(to: address): # -> TwoNums:
+        # (reserve0: u112, reserve1: u112) = self.get_reserves() # gas savings
+        reserve0: u256 = self.reserve0
+        reserve1: u256 = self.reserve1
+        token0: ERC20 = ERC20(self.token0)
+        token1: ERC20 = ERC20(self.token1)
+        balance0: u256 = token0.balanceOf(self.address)
+        balance1: u256 = token1.balanceOf(self.address)
+        liquidity: u256 = self.balances[self.address]
+
+        fee_on: bool = self._mint_fee(reserve0, reserve1)
+        total_supply: u256 = self.total_supply # gas savings, must be defined here since total_supply can update in _mintFee
+        amount0: u256 = (liquidity * balance0) / total_supply # using balances ensures pro-rata distribution
+        amount1: u256 = (liquidity * balance1) / total_supply # using balances ensures pro-rata distribution
+        assert amount0 > 0 and amount1 > 0, "UniswapV2: INSUFFICIENT_LIQUIDITY_BURNED"
+        self._burn(self.address, liquidity)
+        token0.transfer(to, amount0)
+        token1.transfer(to, amount1)
+        balance0 = token0.balanceOf(self.address)
+        balance1 = token1.balanceOf(self.address)
+
+        self._update(balance0, balance1, reserve0, reserve1)
+
+        if fee_on:
+            self.k_last = reserve0 * reserve1 # reserve0 and reserve1 are up-to-date
+
+        emit Burn(msg.sender, amount0, amount1, to)
+        #return TwoNums(num1=amount0, num2=amount1)
+
+    # this low-level function should be called from a contract which performs important safety checks
+    # TODO: add support for the bytes type (https://github.com/ethereum/fe/issues/280)
+    # pub def swap(amount0_out: u256, amount1_out: u256, to: address, data: bytes):
+    pub def swap(amount0_out: u256, amount1_out: u256, to: address):
+        assert amount0_out > 0 or amount1_out > 0, "UniswapV2: INSUFFICIENT_OUTPUT_AMOUNT"
+        reserve0: u256 = self.reserve0
+        reserve1: u256 = self.reserve1
+        assert amount0_out < reserve0 and amount1_out < reserve1, "UniswapV2: INSUFFICIENT_LIQUIDITY"
+
+        token0: ERC20 = ERC20(self.token0)
+        token1: ERC20 = ERC20(self.token1)
+
+        assert to != address(token0) and to != address(token1), "UniswapV2: INVALID_TO"
```

Comment:
We talked about that instead of obtaining the address via `address(token0)` we could also obtain it via `token.address`. I think that would be inline with `self.address`, too. `address` would be a reserved word in the context of a contract properties. Since the ABI only defines functions it would still be allowed for a contract to implement an `address()` function though. We should probably create an issue and put a comment in for that, too.